sled-agent/src/zone_bundle.rs

```diff
@@ -141,6 +161,125 @@ impl ZoneBundleMetadata {
+// there's no long-lived snapshots which accumulate huge deltas from their
+// origins.
+const BASE_DATASET: &'static str = ZONE_BUNDLE_ZFS_DATASET;
+
+// The name of the snapshot (and clone) created from the zone root filesystem.
+const ZONE_ROOT_SNAPSHOT_NAME: &'static str = "zone-root";
+
+// The prefix for all the snapshots (and clones) for each filesystem containing
+// archived logs. Each debug data, such as `oxp_<UUID>/crypt/debug`, generates a
+// snapshot and clone named `zone-archives-<UUID>`.
+const ARCHIVE_SNAPSHOT_PREFIX: &'static str = "zone-archives-";
+
+// An extra ZFS user property attached to all zone bundle snapshots and clones.
+//
+// This is used to ensure that we are not accidentally deleting ZFS objects that
+// a user has created, but which happen to be named the same thing.
+const ZONE_BUNDLE_ZFS_PROPERTY_NAME: &'static str = "for-zone-bundle";
+const ZONE_BUNDLE_ZFS_PROPERTY_VALUE: &'static str = "true";
+
+// Initialize the ZFS resources we need for zone bundling.
+//
+// This first deletes any existing ZFS filesystem that matches the root we use
+// for all zone bundle snapshots and clones. This ensures that a sled-agent
+// crash during zone bundling doesn't leave unwanted snapshots growing without
+// bound as they diverge from their origin.
+//
+// This also recreates that zone-bundle root afterwards.
+#[cfg(not(test))]
+fn initialize_zfs_resources(log: &Logger) -> Result<(), BundleError> {
+    // Destroy the base dataset in which we create any clones during zone
+    // bundling. This will destroy all the clones as well.
+    match Zfs::destroy_dataset(BASE_DATASET) {
+        Ok(_) => {
+            debug!(
+                log,
+                "destroyed pre-existing zone bundle base dataset";
+                "dataset" => BASE_DATASET,
+            );
+        }
+        Err(DestroyDatasetError {
+            err: DestroyDatasetErrorVariant::NotFound,
+            ..
+        }) => {}
+        Err(e) => return Err(BundleError::from(e)),
+    }
+
+    // Delete any existing snapshots that still exist as well.
+    let zb_snapshots =
+        Zfs::list_snapshots().unwrap().into_iter().filter(|snap| {
+            // Check for snapshots named how we expect to create them.
+            if snap.snap_name != ZONE_ROOT_SNAPSHOT_NAME
+                || !snap.snap_name.starts_with(ARCHIVE_SNAPSHOT_PREFIX)
+            {
+                return false;
+            }
+
+            // Additionally check for the zone-bundle-specific property.
+            //
+            // If we find a dataset that matches our names, but which _does not_
+            // have such a property, we'll panic rather than possibly deleting
+            // user data.
+            let name = snap.to_string();
+            let value =
+                Zfs::get_oxide_value(&name, ZONE_BUNDLE_ZFS_PROPERTY_NAME)
+                    .unwrap_or_else(|_| {
+                        panic!(
```

Comment:
`snapshot_and_clone` performs `Zfs::create_snapshot` then `Zfs::set_oxide_value` - if there's a crash in the middle, this would end up stopping the sled agent from booting, right?

Based on the man page, it looks like `zfs snapshot` can accept properties to set:

```
     zfs snapshot [-r] [-o property=value]...
       filesystem@snapname|volume@snapname...
       Creates snapshots with the given names.  All previous modifications by
       successful system calls to the file system are part of the snapshots.
       Snapshots are taken atomically, so that all snapshots correspond to the
       same moment in time.  See the Snapshots section for details.

       -o property=value
           Sets the specified property; see zfs create for details.
```

If it's possible that a crash between those functions could cause the sled-agent to panic when booting, we should set the properties when the snapshot is created.

Reply:
Good idea, thanks. Done in 3ecf04cfc8a4205d23d60f44b5a858306ed6e48e